File: training/t16/H/src/H.java
import java.io.BufferedInputStream;
import java.math.BigInteger;
import java.util.Scanner;

/**
 * Created by cyuun on 2017/7/25.
 */
public class H {
    public static void main(String[] args) {
        Scanner cin = new Scanner(new BufferedInputStream(System.in));
        int t = cin.nextInt();
        for(int k = 1;k<=t; k++)
        {
            BigInteger n,m,rator=BigInteger.ZERO,nator=BigInteger.ZERO;
            n = cin.nextBigInteger();
            m = cin.nextBigInteger();

            for(int i=1;;i++)
            {
                BigInteger compare = BigInteger.valueOf(2*(i-1));
                if(compare.compareTo(n.add(m)) >= 0)break;
                BigInteger up=BigInteger.ONE,down=BigInteger.ONE;
                for(int j=0;j<2*(i-1);j++)
                {
                    up=up.multiply(m.subtract(BigInteger.valueOf(j)));
                    down=down.multiply(n.add(m).subtract(BigInteger.valueOf(j)));
                }
                up=up.multiply(n);down=down.multiply(n.add(m).subtract(compare));
                BigInteger g=up.gcd(down);
                up=up.divide(g);down=down.divide(g);
                //Simplify(up,down);
                if(nator==BigInteger.ZERO&&rator==BigInteger.ZERO) {
                    rator=up;
                    nator=down;
                }
                else {
                    rator=rator.multiply(down).add(nator.multiply(up));
                    nator=nator.multiply(down);
                }
                g=rator.gcd(nator);
                rator=rator.divide(g);nator=nator.divide(g);
                //Simplify(rator,nator);
            }
            System.out.print(rator+"/"+nator+"\n");
        }
    }
}

File: training/t16/H.cpp
#include <bits/stdc++.h>
using namespace std;
typedef unsigned long long ULL;
int t;
ULL rator,nator,up,down,n,m;
void Simplify(ULL &up,ULL &down)
{
    ULL g=__gcd(up,down);
    up/=g;down/=g;
}
int main()
{
    scanf("%d",&t);
    while(t--)
    {
        cin>>n>>m;
        rator=0;nator=0;
        for(int i=1;;i++)
        {
            if(2*(i-1)>=n+m)break;
            up=1;down=1;
            for(int j=0;j<2*(i-1);j++)
            {
                up*=m-j;
                down*=n+m-j;
                Simplify(up,down);
            }
            up*=n;down*=n+m-2*(i-1);
            Simplify(up,down);
            if(nator==0&&rator==0) rator=up,nator=down;
            else rator=rator*down+nator*up,nator=nator*down;
            Simplify(rator,nator);
        }
        cout<<fixed;
        cout<<rator<<"/"<<nator<<endl;
    }
    return 0;
}
